salvo/src/lib/run_benchmark.py

```diff
@@ -0,0 +1,323 @@
+       unavailable.
+
+       Args:
+        images: the DockerImages appearing in the control object
+    """
+    di = docker_image.DockerImage()
+    pull_result = False
+    try:
+      pull_result = di.pull_image(images.nighthawk_benchmark_image)
+    except docker_image.DockerImagePullError:
+      log.error(f"Image pull failed for {images.nighthawk_benchmark_image}")
+
+    if not pull_result:
+      log.debug(f"Attempting to build {images.nighthawk_benchmark_image}")
+      docker_image_builder.build_nighthawk_benchmark_image_from_source(
+          self._source_manager
+      )
+
+  def _pull_or_build_nh_binary_image(self,
+                                     images: proto_image.DockerImages) -> None:
+    """Attempt to pull the NightHawk Binary Image.  Build it if it is
+       unavailable.
+
+      Args:
+        images: the DockerImages appearing in the control object
+    """
+    di = docker_image.DockerImage()
+    pull_result = False
+    try:
+      pull_result = di.pull_image(images.nighthawk_binary_image)
+    except docker_image.DockerImagePullError:
+      log.error(f"Image pull failed for {images.nighthawk_binary_image}")
+
+    if not pull_result:
+      log.debug(f"Attempting to build {images.nighthawk_binary_image}")
+      docker_image_builder.build_nighthawk_binary_image_from_source(
+          self._source_manager
+      )
+
+  def _pull_or_build_nighthawk_images_for_benchmark(self):
+    """Pull the nighthawk docker iamges needed for benchmarks.  If an image
+       remains unavailable, raise an Exception.
+
+    Raises:
+      BenchmarkRunnerError: if no nighthawk images appear in the control object
+    """
+    images = self._control.images
+
+    if not images.nighthawk_benchmark_image:
+      raise BenchmarkRunnerError("No NightHawk Benchmark Image specified")
+
+    if not images.nighthawk_binary_image:
+      raise BenchmarkRunnerError("No NightHawk Binary Image specified")
+
+    # TODO: If bazel options are specified, we need to build the images.
+    # Specific case is if we enable CpuProfiling, this is not available in
+    # the stock image.
+
+    self._pull_or_build_nh_benchmark_image(images)
+    self._pull_or_build_nh_binary_image(images)
+
+  def _pull_or_build_envoy_images_for_benchmark(
+      self, image_hashes: Set[str]) -> Set[str]:
+    """Pull the docker images needed for the benchmarks. If an image is not
+       available build it.
+
+    Args:
+      image_hashes: The envoy image hashes that we are locating images
+
+    Returns:
+      a Set of envoy image tags required for the benchmark:
+        eg ["envoyproxy/envoy:v1.X.X", ...]
+    """
+
+    have_build_options = self._source_manager.have_build_options(
+        proto_source.SourceRepository.SourceIdentity.SRCID_ENVOY)
+
+    envoy_images = set()
+    di = docker_image.DockerImage()
+
+    log.debug(f"Finding matching images for hashes: {image_hashes}")
+
+    for image_hash in image_hashes:
+      assert image_hash
```

Comment:
Assert statements shouldn't be used to validate arguments (I am assuming image_hashes come as an input argument or configuration from the user). Please feel free to correct me if this is purely internal validation.

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#244-decision

Instead we should perform a check and raise an exception explicitly.

Reply:
This can be removed.  Likely a leftover artifact from debugging.  I haven't seen this assert fail.